components/provider/src/resource.rs

```diff
@@ -0,0 +1,392 @@
+///
+/// The fields in a `ResourceOptions` are not generally known until runtime.
+#[derive(PartialEq, Clone)]
+pub struct ResourceOptions {
+    // TODO: Consider making multiple variant fields.
+    pub variant: Option<Cow<'static, str>>,
+    pub langid: Option<LanguageIdentifier>,
+}
+
+impl fmt::Debug for ResourceOptions {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "ResourceOptions{{{}}}", self)
+    }
+}
+
+impl fmt::Display for ResourceOptions {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        let components = self.get_components();
+        let mut it = components.iter();
+        if let Some(s) = it.next() {
+            f.write_str(s)?;
+            for s in it {
+                f.write_char('/')?;
+                f.write_str(s)?;
+            }
+        }
+        Ok(())
+    }
+}
+
+impl Default for ResourceOptions {
+    fn default() -> Self {
+        Self {
+            variant: None,
+            langid: None,
+        }
+    }
+}
+
+impl ResourceOptions {
+    /// Gets the standard path components of this `ResourceOptions`. These components should be used when
+    /// persisting the `ResourceOptions` on the filesystem or in structured data.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// use std::borrow::Cow;
+    /// use icu_provider::prelude::*;
+    /// use icu_locid_macros::langid;
+    ///
+    /// let resc_options = ResourceOptions {
+    ///     variant: Some(Cow::Borrowed("GBP")),
+    ///     langid: Some(langid!("pt_BR")),
+    /// };
+    /// let components = resc_options.get_components();
+    ///
+    /// assert_eq!(
+    ///     ["GBP", "pt-BR"],
+    ///     components.iter().collect::<Vec<&str>>()[..]
+    /// );
+    /// ```
+    pub fn get_components(&self) -> ResourceOptionsComponents {
+        self.into()
+    }
+
+    /// Returns whether this `ResourceOptions` has all empty fields (no components).
+    pub fn is_empty(&self) -> bool {
+        self == &Self::default()
+    }
+}
+
+/// The standard components of a ResourceOptions path.
+pub struct ResourceOptionsComponents {
+    components: [Option<Cow<'static, str>>; 2],
+}
+
+impl ResourceOptionsComponents {
+    pub fn iter(&self) -> impl Iterator<Item = &str> {
+        self.components
+            .iter()
+            .filter_map(|option| option.as_ref().map(|cow| cow.borrow()))
+    }
+}
+
+impl From<&ResourceOptions> for ResourceOptionsComponents {
+    fn from(resc_options: &ResourceOptions) -> Self {
+        Self {
+            components: [
+                resc_options.variant.as_ref().cloned(),
+                resc_options
+                    .langid
+                    .as_ref()
+                    .map(|s| Cow::Owned(s.to_string())),
+            ],
+        }
+    }
+}
+
+#[derive(Clone, Debug, PartialEq)]
+pub struct ResourcePath {
+    pub key: ResourceKey,
+    pub options: ResourceOptions,
+}
+
+#[test]
+fn test_entry_to_string() {
```

Comment:
in-source tests are usually wrapped in `#[cfg(test])]` `test` module - https://doc.rust-lang.org/rust-by-example/testing/unit_testing.html

Reply:
This is new to me!  I swear I don't remember seeing this in Rust By Example when I read it 1.5 years ago.  I'll make this style change here and adjust it elsewhere when it comes up.